### DISCUSSION TITLE
Limitations of Native player vs. Swiftfin (VLCKit)

### DISCUSSION THREAD

**Discussion by holow29** [jellyfin/Swiftfin#953] [Discussions]
There is more information in #519, but for the sake of consolidation, I thought I would put some points of interest here in terms of comparing the Native Apple AVPlayer implementation with that of Swiftfin's VLCKit implementation. This might be hepful when MPV is implemented to look at how this changes.

The main question is why one might choose the Native player over Swiftfin VLCKit or vice versa.

| Feature | Swiftfin/VLCKit  | Native |
| ------------- | ------------- | ------------- |
| Framerate matching  | | ✓ |
| HDR playback | | HDR10, HDR10+*, Dolby Vision Profiles 5 & 8 |
| HDR->SDR tonemapping | ? Will play but unsure if in correct colorspace | If DirectPlaying MP4 |
| Subtitle support | Extensive, incl. image-based (PGS, etc.) | Extremely limited, text-based (WebVTT) |
| Container/codec support | Extensive, will DirectPlay almost everything (except TrueHD/Atmos/MLP) | Extremely limited, modern codecs will generally remux, others transcode |

Given the versatility of Jellyfin, many of these limitations can be overcome. For example, Jellyfin can transcode audio/video to suit what the player requests, including HDR->SDR tonemapping. It can also take text-based subtitles and convert them to other text-based subs like SRT to WebVTT, or it can burn-in image-based subs to the video stream. However, the goal for people is usually to avoid transcoding, so knowing these limitations might aid in that.

(This does not take into account the interface differences right now; for example, switching subtitles/audio streams is not working correctly on Native player or on Swiftfin VLCKit when transcoding.)

My general advice is the following:
Use native if all are true:
- Content is in AVC or HEVC
- You have an HDR display or content is not HDR
- You don't need subtitle support

Otherwise, use Swiftfin VLCKit.

**Comment by JPKribs** [comment]
Hey @holow29! I’m a big fan of this chart and thought it would be great to make it a more permanent fixture. I’ve opened a PR to add a dedicated page for it here: [#1302](https://github.com/jellyfin/Swiftfin/pull/1302).

If you have a chance, I’d really appreciate a second set of eyes to check for any missing or inaccurate information. Let me know if you have any notes!

**Comment by JPKribs** [comment]
Closing out in favor our the new documentation. Thank you for your help getting this together!

New Documentation can be found here: https://github.com/jellyfin/Swiftfin/blob/main/Documentation/players.md
